### PR TITLE
[clang][OpenMP] Slightly refactor EndOpenMPDSABlock for readability, NFC

### PR DESCRIPTION
**Pull request by kparzysz**
Change the loop
```
  if (isOpenMPExecutableDirective)
    for (Clause)
      if (Clause is kind1)
        multi
        line
        do
        something1;
      else if (Clause is kind2)
        ...
      ...
```
to
```
  auto do1 = ...do something1...;
  auto do2 = ...do something2...;
  ...

  if (isOpenMPExecutableDirective)
    for (Clause)
      if (Clause is kind1)
        do1();
      else if (Clause is kind2)
        do2();
      ...
    ...
```

### CODE REVIEW

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 1]
```diff
@@ -2861,113 +2861,120 @@ void SemaOpenMP::EndOpenMPDSABlock(Stmt *CurDirective) {
...
-          if (!D)
-            continue;
 
-          const DSAStackTy::DSAVarData DVar =
-              DSAStack->getTopDSA(D, /*FromParent=*/false);
-          PrivateRefs.push_back(DVar.PrivateCopy ? DVar.PrivateCopy
-                                                 : SimpleRefExpr);
-        }
-        Clause->setPrivateRefs(PrivateRefs);
+  auto finalizeLastprivate = [&](OMPLastprivateClause *Clause) {
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:2865]: FinalizeLastprivate

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 2]
```diff
@@ -2861,113 +2861,120 @@ void SemaOpenMP::EndOpenMPDSABlock(Stmt *CurDirective) {
...
-            else
-              Diag(MapExpr->getExprLoc(), diag::note_used_here)
-                  << MapExpr->getSourceRange();
-          }
+      auto *DRE = cast<DeclRefExpr>(DE->IgnoreParens());
+      auto *VD = cast<VarDecl>(DRE->getDecl());
+      QualType Type = VD->getType().getNonReferenceType();
+      const DSAStackTy::DSAVarData DVar =
+          DSAStack->getTopDSA(VD, /*FromParent=*/false);
+      if (DVar.CKind == OMPC_lastprivate) {
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:2877]: Use early exit where possible

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 3]
```diff
@@ -2861,113 +2861,120 @@ void SemaOpenMP::EndOpenMPDSABlock(Stmt *CurDirective) {
...
+      } else {
+        // The variable is also a firstprivate, so initialization sequence
+        // for private copy is generated already.
+        PrivateCopies.push_back(nullptr);
+      }
+    }
+    Clause->setPrivateCopies(PrivateCopies);
+  };
+
+  auto finalizeNontemporal = [&](OMPNontemporalClause *Clause) {
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:2902]: FinalizeNontemporal

File: clang/lib/Sema/SemaOpenMP.cpp [hunk 4]
```diff
@@ -2861,113 +2861,120 @@ void SemaOpenMP::EndOpenMPDSABlock(Stmt *CurDirective) {
...
+
+      const DSAStackTy::DSAVarData DVar =
+          DSAStack->getTopDSA(D, /*FromParent=*/false);
+      PrivateRefs.push_back(DVar.PrivateCopy ? DVar.PrivateCopy
+                                             : SimpleRefExpr);
+    }
+    Clause->setPrivateRefs(PrivateRefs);
+  };
+
+  auto finalizeAllocators = [&](OMPUsesAllocatorsClause *Clause) {
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenMP.cpp:2926]: FinalizeAllocators

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 1 above)
Comment (kparzysz) [reply clang/lib/Sema/SemaOpenMP.cpp]: Done.

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 2 above)
Comment (kparzysz) [reply clang/lib/Sema/SemaOpenMP.cpp]: Done.

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 3 above)
Comment (kparzysz) [reply clang/lib/Sema/SemaOpenMP.cpp]: Done.

File: clang/lib/Sema/SemaOpenMP.cpp (hunk 4 above)
Comment (kparzysz) [reply clang/lib/Sema/SemaOpenMP.cpp]: Done.

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**